File: 813-AllPathsFromSourceToTarget/813-AllPathsFromSourceToTarget.js
// Last updated: 4/29/2026, 10:27:34 AM
/**
 * @param {number[][]} graph
 * @return {number[][]}
 */
// bfs
// 2022-11-13
// 05:00s
// O(n)
// O(1)
var allPathsSourceTarget = function(graph) {
    const n = graph.length;
    const ans = [];
    
    const bfs = path => {
        const last = path[path.length - 1];
        if(last === n - 1) {
            ans.push(path);
            return;
        }
        
        const neighbors = graph[last];
        for(const neighbor of neighbors) {
            bfs(path.concat([neighbor]))
        }
    }
    
    bfs([0]);
    return ans;
};

// 2022-11-13
// 30:00s
// think
// think
var allPathsSourceTarget = function(graph) {
    const n = graph.length;
    const ans = [];
    
    const backtrack = path => {
        // console.log(path);
        const last = path[path.length - 1];
        // console.log(`last: ${last}`);
        if(last === n - 1) {
            ans.push([...path]);
            return;
        }
        
        const neighbors = graph[last];
        // console.log(`neighbors: ${neighbors}`)
        for(const neighbor of neighbors) {
            path.push(neighbor);
            backtrack(path);
            path.pop();
        }
    }
    
    backtrack([0]);
    return ans;
};

File: 813-AllPathsFromSourceToTarget/813-AllPathsFromSourceToTarget.py
# Last updated: 4/29/2026, 10:27:39 AM
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        n = len(graph)
        path_list = []
        
        def dfs(start, final_end, path, visited, l):
            if start == final_end:
                l.append(path)
                return
            visited.add(start)
            for end in graph[start]:
                if not end in visited:
                    new_path = path + [end]
                    dfs(end, final_end, new_path, visited.copy(), l)
            
        
        dfs(0, n-1, [0], set(), path_list);
        return path_list
    
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        n = len(graph)
        path_list = []
        
        def backtrack(start, final_end, path, visited, l):
            if start == final_end:
                l.append(path)
                return
            visited.add(start)
            for end in graph[start]:
                if not end in visited:
                    new_path = path + [end]
                    backtrack(end, final_end, new_path, visited, l)
            path.pop()
            visited.remove(start)
            
        
        backtrack(0, n-1, [0], set(), path_list);
        return path_list
    
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        target = len(graph) - 1
        res = []
        
        def backtrack(curr_node, path, visited):
            if curr_node == target:
                res.append(list(path))
                return
            visited.add(curr_node)
            for next_node in graph[curr_node]:
                if not next_node in visited:
                    path.append(next_node)
                    backtrack(next_node, path, visited)
                    path.pop()
            visited.remove(curr_node)
            
        
        backtrack(0, [0], set());
        return res